internal/controller/thanosruler_controller.go

```diff
@@ -0,0 +1,383 @@
+		if err != nil {
+			r.logger.Error(
+				err, "failed to create or update resource",
+				"gvk", obj.GetObjectKind().GroupVersionKind().String(),
+				"resource", obj.GetName(),
+				"namespace", obj.GetNamespace(),
+			)
+			errCount++
+			continue
+		}
+
+		r.logger.V(1).Info(
+			"resource configured",
+			"operation", op, "gvk", obj.GetObjectKind().GroupVersionKind().String(),
+			"resource", obj.GetName(), "namespace", obj.GetNamespace(),
+		)
+	}
+
+	if errCount > 0 {
+		r.ControllerBaseMetrics.ClientErrorsTotal.WithLabelValues(manifestruler.Name).Add(float64(errCount))
+		return fmt.Errorf("failed to create or update %d resources for the ruler", errCount)
+	}
+
+	return nil
+}
+
+func (r *ThanosRulerReconciler) buildRuler(ctx context.Context, ruler monitoringthanosiov1alpha1.ThanosRuler) []client.Object {
+	metaOpts := manifests.Options{
+		Name:      ruler.GetName(),
+		Namespace: ruler.GetNamespace(),
+		Replicas:  ruler.Spec.Replicas,
+		Labels:    ruler.GetLabels(),
+		Image:     ruler.Spec.Image,
+		LogLevel:  ruler.Spec.LogLevel,
+		LogFormat: ruler.Spec.LogFormat,
+	}.ApplyDefaults()
+
+	endpoints := r.getQueryAPIServiceEndpoints(ctx, ruler)
+	ruleFiles := r.getRuleConfigMaps(ctx, ruler)
+
+	additional := manifests.Additional{
+		Args:         ruler.Spec.Additional.Args,
+		Containers:   ruler.Spec.Additional.Containers,
+		Volumes:      ruler.Spec.Additional.Volumes,
+		VolumeMounts: ruler.Spec.Additional.VolumeMounts,
+		Ports:        ruler.Spec.Additional.Ports,
+		Env:          ruler.Spec.Additional.Env,
+		ServicePorts: ruler.Spec.Additional.ServicePorts,
+	}
+	return manifestruler.BuildRuler(manifestruler.RulerOptions{
+		Options:         metaOpts,
+		Endpoints:       endpoints,
+		RuleFiles:       ruleFiles,
+		ObjStoreSecret:  ruler.Spec.ObjectStorageConfig.ToSecretKeySelector(),
+		Retention:       manifests.Duration(ruler.Spec.Retention),
+		AlertmanagerURL: ruler.Spec.AlertmanagerURL,
+		ExternalLabels:  ruler.Spec.ExternalLabels,
+		AlertLabelDrop:  ruler.Spec.AlertLabelDrop,
+		StorageSize:     resource.MustParse(ruler.Spec.StorageSize),
+		Additional:      additional,
+	})
+}
+
+// getStoreAPIServiceEndpoints returns the list of endpoints for the QueryAPI services that match the ThanosRuler queryLabelSelector.
+func (r *ThanosRulerReconciler) getQueryAPIServiceEndpoints(ctx context.Context, ruler monitoringthanosiov1alpha1.ThanosRuler) []manifestruler.Endpoint {
+	services := &corev1.ServiceList{}
+	if err := r.List(
+		ctx,
+		services,
+		[]client.ListOption{
+			client.MatchingLabels(ruler.Spec.QueryLabelSelector.MatchLabels),
+			client.InNamespace(ruler.Namespace),
+		}...); err != nil {
+		return []manifestruler.Endpoint{}
+	}
+
+	if len(services.Items) == 0 {
+		return []manifestruler.Endpoint{}
+	}
+
+	endpoints := make([]manifestruler.Endpoint, len(services.Items))
+	for i, svc := range services.Items {
+		for _, port := range svc.Spec.Ports {
+			if port.Name == "grpc" {
```

Comment:
prefer to use https://github.com/thanos-community/thanos-operator/blob/main/internal/pkg/manifests/query/builder.go#L25 here if we can. 

We probably also want to document this behaviour in general (outside of this PR and across the board for all components!)